File: scripts/install/src/cq_install/hosts/cursor.py
"""Cursor host adapter."""

from __future__ import annotations

import hashlib
import platform
import shlex
import subprocess
from pathlib import Path

from cq_install.common import (
    copy_selected_paths,
    copy_tree,
    remove_copied_tree,
    remove_hook_entry,
    remove_json_entry,
    remove_owned_file,
    upsert_hook_entry,
    upsert_json_entry,
    write_if_missing,
)
from cq_install.content import (
    CQ_MCP_KEY,
    CQ_RUNTIME_MANIFEST,
    PYTHON_COMMAND,
    cq_binary_name,
)
from cq_install.context import ChangeResult, InstallContext
from cq_install.hosts.base import HostDef
from cq_install.runtime import runtime_root

CURSOR_RULE_CONTENT = """---
description: cq shared knowledge commons
alwaysApply: true
---

Before starting any implementation task, load the `cq` skill and follow its Core Protocol.
"""

CURSOR_HOOK_MODES: list[tuple[str, str]] = [
    ("sessionStart", "session-start"),
    ("postToolUseFailure", "post-tool-use-failure"),
    ("postToolUse", "post-tool-use"),
    ("stop", "stop"),
]

# Populate when the hook command shape changes in a future release.
# Entries matching these strings are removed on install before the desired
# entry is (re)inserted, giving a clean migration path for stale commands.
CURSOR_LEGACY_HOOK_COMMANDS: list[str] = []

CURSOR_HOOKS_FILE = "hooks.json"
CURSOR_HOOK_SCRIPT_RELPATH = Path("hooks") / "cursor" / "cq_cursor_hook.py"
CURSOR_HOST_SKILLS_MANIFEST = ".cq-install-manifest.json"
CURSOR_MCP_FILE = "mcp.json"
CURSOR_MCP_SERVERS_KEY = "mcpServers"
CURSOR_RULE_HASH = hashlib.sha256(CURSOR_RULE_CONTENT.encode()).hexdigest()
CURSOR_RULE_RELPATH = Path("rules") / "cq.mdc"
CURSOR_SKILLS_DIR = "skills"
CURSOR_STATE_DIR = "cq-hook-state"

# Cursor reads its config from ~/.cursor on every platform and from
# <project>/.cursor for per-project installs.
CURSOR_TARGET_DIR = ".cursor"


class CursorHost(HostDef):
    """Adapter for the Cursor editor host."""

    name = "cursor"

    def global_target(self) -> Path:
        """Return the global Cursor config dir."""
        return Path.home() / CURSOR_TARGET_DIR

    def install(self, ctx: InstallContext) -> list[ChangeResult]:
        """Install cq into the Cursor target."""
        results: list[ChangeResult] = []
        results.extend(self._install_skills(ctx))
        results.extend(ctx.run_state.ensure_cq_binary(ctx))
        results.append(self._install_runtime(ctx))
        results.append(self._install_mcp(ctx))
        results.append(
            write_if_missing(
                ctx.target / CURSOR_RULE_RELPATH,
                CURSOR_RULE_CONTENT,
                dry_run=ctx.dry_run,
            )
        )
        results.extend(self._install_hooks(ctx))
        return results

    def project_target(self, project: Path) -> Path:
        """Return the per-project Cursor config dir."""
        return project / CURSOR_TARGET_DIR

    def uninstall(self, ctx: InstallContext) -> list[ChangeResult]:
        """Remove cq from the Cursor target."""
        results: list[ChangeResult] = [
            remove_copied_tree(
                ctx.target / CURSOR_SKILLS_DIR,
                manifest_name=CURSOR_HOST_SKILLS_MANIFEST,
                dry_run=ctx.dry_run,
            ),
            remove_json_entry(
                ctx.target / CURSOR_MCP_FILE,
                [CURSOR_MCP_SERVERS_KEY, CQ_MCP_KEY],
                dry_run=ctx.dry_run,
            ),
            remove_owned_file(
                ctx.target / CURSOR_RULE_RELPATH,
                expected_content_hash=CURSOR_RULE_HASH,
                dry_run=ctx.dry_run,
            ),
        ]
        for hook_name, mode in CURSOR_HOOK_MODES:
            results.append(
                remove_hook_entry(
                    ctx.target / CURSOR_HOOKS_FILE,
                    hook_name=hook_name,
                    command=_hook_command(ctx, mode),
                    dry_run=ctx.dry_run,
                )
            )
        return results

    def _install_hooks(self, ctx: InstallContext) -> list[ChangeResult]:
        results: list[ChangeResult] = []
        for hook_name, mode in CURSOR_HOOK_MODES:
            results.append(
                upsert_hook_entry(
                    ctx.target / CURSOR_HOOKS_FILE,
                    hook_name=hook_name,
                    command=_hook_command(ctx, mode),
                    legacy_commands=CURSOR_LEGACY_HOOK_COMMANDS,
                    dry_run=ctx.dry_run,
                )
            )
        return results

    def _install_mcp(self, ctx: InstallContext) -> ChangeResult:
        binary_path = runtime_root() / "bin" / cq_binary_name()
        return upsert_json_entry(
            ctx.target / CURSOR_MCP_FILE,
            [CURSOR_MCP_SERVERS_KEY, CQ_MCP_KEY],
            {
                "command": str(binary_path),
                "args": ["mcp"],
            },
            dry_run=ctx.dry_run,
        )

    def _install_runtime(self, ctx: InstallContext) -> ChangeResult:
        return copy_selected_paths(
            ctx.plugin_root,
            runtime_root(),
            relpaths=[CURSOR_HOOK_SCRIPT_RELPATH],
            manifest_name=CQ_RUNTIME_MANIFEST,
            dry_run=ctx.dry_run,
        )

    def _install_skills(self, ctx: InstallContext) -> list[ChangeResult]:
        if ctx.host_isolated_skills:
            return [
                copy_tree(
                    ctx.plugin_root / CURSOR_SKILLS_DIR,
                    ctx.target / CURSOR_SKILLS_DIR,
                    manifest_name=CURSOR_HOST_SKILLS_MANIFEST,
                    dry_run=ctx.dry_run,
                )
            ]
        return ctx.run_state.ensure_shared_skills(ctx)


def _hook_command(ctx: InstallContext, mode: str) -> str:
    # Cursor's hooks.json takes a single shell-executable string per hook
    # entry. POSIX shell and Windows cmd.exe have different quoting rules,
    # so we pick the right stdlib helper at install time. See also
    # cursor/cursor#3386 for a related Cursor-side path-quoting bug on
    # Windows; worth checking if hook commands misbehave there.
    hook_script = runtime_root() / CURSOR_HOOK_SCRIPT_RELPATH
    state_dir = ctx.target / CURSOR_STATE_DIR
    parts = [
        PYTHON_COMMAND,
        str(hook_script),
        "--mode",
        mode,
        "--state-dir",
        str(state_dir),
    ]
    if platform.system() == "Windows":
        return subprocess.list2cmdline(parts)
    return shlex.join(parts)

File: cli/internal/install/cursor.go
package install

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/mozilla-ai/cq/sdk/go/prompts"

	"github.com/mozilla-ai/cq/cli/internal/hook"
)

const (
	// cursorMCPFile is the MCP configuration file Cursor reads.
	cursorMCPFile = "mcp.json"

	// cursorHooksFile is the lifecycle-hook configuration file Cursor reads.
	cursorHooksFile = "hooks.json"

	// cursorStateDir holds the cq hook's cross-event state within the target.
	cursorStateDir = "cq-hook-state"

	// cursorRule is the always-applied Cursor rule pointing agents at the cq skill.
	cursorRule = "---\n" +
		"description: cq shared knowledge commons\n" +
		"alwaysApply: true\n" +
		"---\n\n" +
		"Before starting any implementation task, load the `cq` skill and follow its Core Protocol.\n"
)

// cursorRuleRelPath is the always-applied rule file, relative to the target.
var cursorRuleRelPath = filepath.Join("rules", "cq.mdc")

// cursorHookModes maps each Cursor lifecycle event (the hooks.json key) to the
// canonical cq hook mode passed to `cq _hook --mode`.
var cursorHookModes = []struct {
	// event is the Cursor hooks.json key for the lifecycle event.
	event string

	// mode is the canonical cq mode passed to cq _hook --mode.
	mode hook.Mode
}{
	{"sessionStart", hook.ModeSessionStart},
	{"postToolUseFailure", hook.ModePostToolUseFailure},
	{"postToolUse", hook.ModePostToolUse},
	{"stop", hook.ModeStop},
}

// cursorHost installs cq into the Cursor editor: the shared skill, the MCP
// server entry, an always-applied rule, and the lifecycle hooks.
//
// Cursor stores its config under <home>/.cursor on every platform and is
// global-only; it reads skills from the shared commons.
type cursorHost struct{}

// GlobalTarget returns the Cursor config dir under home.
func (cursorHost) GlobalTarget(home string) string {
	return cursorTarget(home)
}

// Install writes the shared skill, the MCP entry, the rule, and the hooks.
func (cursorHost) Install(ctx Context) ([]Change, error) {
	skill, err := writeManagedFiles(ctx.SkillsDir, map[string]string{
		filepath.Join("cq", "SKILL.md"): prompts.Skill(),
	}, ctx.DryRun)
	if err != nil {
		return nil, err
	}
	changes := []Change{skill}

	mcp, err := upsertJSONEntry(
		filepath.Join(ctx.Target, cursorMCPFile),
		[]string{"mcpServers", "cq"},
		map[string]any{"command": ctx.BinaryPath, "args": []any{"mcp"}},
		ctx.DryRun,
	)
	if err != nil {
		return nil, err
	}
	changes = append(changes, mcp)

	rule, err := writeIfMissing(filepath.Join(ctx.Target, cursorRuleRelPath), cursorRule, ctx.DryRun)
	if err != nil {
		return nil, err
	}
	changes = append(changes, rule)

	for _, hm := range cursorHookModes {
		h, err := upsertHookEntry(
			filepath.Join(ctx.Target, cursorHooksFile),
			hm.event,
			cursorHookCommand(ctx, hm.mode),
			ctx.DryRun,
		)
		if err != nil {
			return nil, err
		}
		changes = append(changes, h)
	}
	return changes, nil
}

// Name returns the host identifier.
func (cursorHost) Name() Target { return TargetCursor }

// SupportsProject reports that Cursor is installed globally in this phase.
func (cursorHost) SupportsProject() bool { return false }

// Uninstall removes the MCP entry, the rule (when unmodified), the hook
// entries, and the hook state dir.
//
// NOTE: the skill lives in the shared commons (~/.agents/skills), which other
// hosts may also use, so it is intentionally left in place.
func (cursorHost) Uninstall(ctx Context) ([]Change, error) {
	mcp, err := removeJSONEntry(
		filepath.Join(ctx.Target, cursorMCPFile),
		[]string{"mcpServers", "cq"},
		ctx.DryRun,
	)
	if err != nil {
		return nil, err
	}
	changes := []Change{mcp}

	rule, err := removeOwnedFile(filepath.Join(ctx.Target, cursorRuleRelPath), sha256Hex(cursorRule), ctx.DryRun)
	if err != nil {
		return nil, err
	}
	changes = append(changes, rule)

	for _, hm := range cursorHookModes {
		h, err := removeHookEntry(
			filepath.Join(ctx.Target, cursorHooksFile),
			hm.event,
			cursorHookCommand(ctx, hm.mode),
			ctx.DryRun,
		)
		if err != nil {
			return nil, err
		}
		changes = append(changes, h)
	}

	stateChange, err := removeCursorStateDir(ctx)
	if err != nil {
		return nil, err
	}
	changes = append(changes, stateChange)
	return changes, nil
}

// cursorHookCommand builds the shell command Cursor runs for one hook event:
// this binary's hidden _hook handler, with the resolved state dir.
func cursorHookCommand(ctx Context, mode hook.Mode) string {
	return shellJoin(
		ctx.BinaryPath,
		"_hook",
		"--host", string(hook.HostCursor),
		"--mode", string(mode),
		"--state-dir", filepath.Join(ctx.Target, cursorStateDir),
	)
}

// removeCursorStateDir deletes the hook state dir, which the installer and the
// cq hook own entirely (no user files are ever written there).
func removeCursorStateDir(ctx Context) (Change, error) {
	dir := filepath.Join(ctx.Target, cursorStateDir)
	if _, err := os.Stat(dir); os.IsNotExist(err) {
		return Change{Action: ActionUnchanged, Path: dir}, nil
	} else if err != nil {
		return Change{}, fmt.Errorf("checking hook state directory: %w", err)
	}
	if !ctx.DryRun {
		if err := os.RemoveAll(dir); err != nil {
			return Change{}, fmt.Errorf("removing hook state directory: %w", err)
		}
	}
	return Change{Action: ActionRemoved, Path: dir}, nil
}
